### PR TITLE
Ensure that all rewriteable are called in retrievers

### PR DESCRIPTION
**Pull request by jimczi**
This PR ensures that all retriever applies the rewrite to all their rewriteable. Rewriting eagerly at the retriever level ensures that we don't rewrite the same query multiple times when compound retrievers are used.
This is marked as non-issue since this is an unreleased bug.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/retriever/KnnRetrieverBuilder.java [hunk 1]
```diff
@@ -111,20 +115,64 @@ public KnnRetrieverBuilder(
         Float similarity
     ) {
         this.field = field;
-        this.queryVector = queryVector;
+        this.queryVector = () -> queryVector;
```
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/search/retriever/KnnRetrieverBuilder.java:118]: Maybe we could also add a validation that only one of `queryVector` or `queryVectorBuilder` can be provided during parsing?

File: server/src/main/java/org/elasticsearch/search/retriever/KnnRetrieverBuilder.java [hunk 2]
```diff
@@ -111,20 +115,64 @@ public KnnRetrieverBuilder(
...
+                        );
+                        return;
+                    }
+                    ll.onResponse(null);
+                }));
+            });
+            var rewritten = new KnnRetrieverBuilder(this, () -> toSet.get(), null);
+            return rewritten;
+        }
+        return super.rewrite(ctx);
```
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/search/retriever/KnnRetrieverBuilder.java:173]: Should we make the parent `RetrieverBuilder#rewrite` abstract to ensure that all retrievers have to implement rewrite?

**Review by pmpailis — APPROVED** [review]
LGTM!

File: server/src/main/java/org/elasticsearch/search/retriever/KnnRetrieverBuilder.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/retriever/KnnRetrieverBuilder.java]: Addressed in https://github.com/elastic/elasticsearch/pull/114366/commits/617dff766d6a9cfd19ab6dd5a13c080e66d30299

**Pull request merged**